src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Registration/RegistrationValidator.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.Services.Metadata.Catalog.Monitoring
+{
+    public abstract class RegistrationValidator : Validator<RegistrationEndpoint>
+    {
+        private IPackageRegistrationMetadataResource V2Resource;
+        private IPackageRegistrationMetadataResource V3Resource;
+
+        public RegistrationValidator(
+            IDictionary<FeedType, SourceRepository> feedToSource, 
+            ILogger<RegistrationValidator> logger) : base(feedToSource, logger)
+        {
+            V2Resource = feedToSource[FeedType.HttpV2].GetResource<IPackageRegistrationMetadataResource>();
+            V3Resource = feedToSource[FeedType.HttpV3].GetResource<IPackageRegistrationMetadataResource>();
+        }
+
+        protected override async Task RunInternal(ValidationContext data)
+        {
+            var exceptions = new List<Exception>();
+
+            try
+            {
+                await CompareIndex(
+                    await GetIndex(V2Resource, data),
+                    await GetIndex(V3Resource, data));
+            }
+            catch (Exception e)
+            {
+                exceptions.Add(new ValidationException("Registration index metadata does not match the FindPackagesById metadata!", e));
+            }
+
+            try
+            {
+                await CompareLeaf(
+                    await GetLeaf(V2Resource, data),
+                    await GetLeaf(V3Resource, data));
+            }
+            catch (Exception e)
+            {
+                exceptions.Add(new ValidationException("Registration leaf metadata does not match the Packages(Id='...',Version='...') metadata!", e));
+            }
+
+            if (exceptions.Any())
+            {
+                throw new AggregateException(exceptions);
+            }
+        }
+
+        public abstract Task CompareIndex(PackageRegistrationIndexMetadata v2, PackageRegistrationIndexMetadata v3);
```

Comment:
Why do `CompareIndex` and `CompareLeaf` not have `async` on their method signatures? Is that intentional?